torch/fx/proxy.py

```diff
@@ -1,17 +1,83 @@
 import dis
+import copy
 import torch
 import inspect
 import operator
 import traceback
+import collections
 
 from .graph import magic_methods, reflectable_magic_methods, Graph
-from typing import Tuple, Dict, Optional, Iterable, Any, Iterator, Callable
+from typing import Tuple, Dict, OrderedDict, Optional, Iterable, Any, Iterator, Callable
 from .node import Target, Node, Argument, base_types, map_aggregate
 from ._compatibility import compatibility
 from .operator_schemas import check_for_mutable_operation
 import torch.fx.traceback as fx_traceback
 
-__all__ = ['TracerBase', 'GraphAppendingTracer', 'TraceError', 'Proxy', 'Attribute', 'ParameterProxy']
+__all__ = ['TracerBase', 'GraphAppendingTracer', 'TraceError',
+           'Proxy', 'Attribute', 'ParameterProxy', 'Scope',
+           'ScopeContextManager']
+
+
+@compatibility(is_backward_compatible=False)
+class Scope(object):
+    """ Scope object that records the module path and the module type
+    of a module. Scope is used to track the information of the module
+    that contains a Node in a Graph of GraphModule. For example::
+
+        class Sub(torch.nn.Module):
+            def forward(self, x):
+                # This will be a call_method Node in GraphModule,
+                # scope for this would be (module_path="sub", module_type=Sub)
+                return x.transpose(1, 2)
+
+        class M(torch.nn.Module):
+            def __init__(self):
+                self.sub = Sub()
+
+            def forward(self, x):
+                # This will be a call_method Node as well,
+                # scope for this would be (module_path="", None)
+                x = x.transpose(1, 2)
+                x = self.sub(x)
+                return x
+
+    """
+
+    def __init__(self, module_path: str, module_type: Any):
+        super().__init__()
+        self.module_path = module_path
+        self.module_type = module_type
+
+
+@compatibility(is_backward_compatible=False)
+class ScopeContextManager(object):
+    """ A context manager to track the Scope of Node during symbolic tracing.
+    When entering a forward function of a Module, we'll update the scope information of
+    the current module, and when we exit, we'll restore the previous scope information.
+    """
+
+    def __init__(
+        self,
+        prev_scope: Scope,
+        current_scope: Scope,
+    ):
+        super().__init__()
+        # Make a copy of scope to restore on context block exit
+        self._prev_scope = copy.copy(prev_scope)
+        # Modify the prev_scope that was passed in by reference
+        prev_scope.module_path = current_scope.module_path
+        prev_scope.module_type = current_scope.module_type
```

Comment:
why do we have to do this?

Reply:
OK I looked again, it looks like `prev_scope` is actually a `scope` variable that keeps track of current scope, maybe we should update the name to reflect this better, e.g. update `prev_scope` to just `scope`

Reply:
I feel something like the following might be clearer (pretty similar to original code)
```
def __init__(
        self,
        scope: Scope,
        current_scope: Scope,
    ):
    # keep a copy of previous scope so that we can restore the scope
    self._prev_scope = copy.copy(scope)
    # update scope to current scope
    scope.module_path = current_scope.module_path
    scope.module_type = current_scope.module_type
    # holding a reference so that we can restore it
    self.scope = scope

...

def __exit__(self, *args):
        # restore the scope to previous scope
        self.scope.module_path = self._prev_scope.module_path
        self.scope.module_type = self._prev_scope.module_type
        return
```

Reply:
You beat me to it.  Your comments above make sense.  Using the name prev_scope made the code confusing.